### PR TITLE
feat(directory): molgenis-emx2-directory-client

### PR DESCRIPTION
**Pull request by dtroelofsprins**
What are the main changes you did:
Created a new tool: molgenis-emx2-directory-client, which is also known as the EMX2 Publish script for BBMRI-ERIC, which combines data from the staging areas, add some extra fields and adds the data to the ERIC tables which are used in the Directory App. 
This is initial a copy of molgenis-py-bbmri-eric 1.18.1, which is the EMX1 version of the Publish script.

how to test:
- On a test-server, create a database based on the biobank_directory template (no demo data)
- Create one or more databases based on the biobank_directory_staging template (with demo data)
- Create a script by copy paste the "dev.py"
- Add the latest version on testPyPI of molgenis-emx2-directory-client to the requirements
- Run the script
- Check that the data from the staging area(s) is in the directory database 

todo:
- [x] Adjust the project setup files to get it working in EMX2
~~- [ ] Adjust README~~ => separate [ticket](https://github.com/molgenis/GCC/issues/820)
- [x] Replace molgenis-py-client with molgenis-emx2-pyclient
~~- [ ] updated docs in case of new feature~~ => separate [ticket](https://github.com/molgenis/GCC/issues/820) 
- [x] updated tests

~~Optional todo:~~ separate [ticket](https://github.com/molgenis/GCC/issues/820)
- [x] test coverage is not 76%, but is not checked before final merge, anyway perhaps some extra tests could be added
- [x] Refactoring: Some code has become redundant (f.e. convert_to_staging as tables names are now equal between ERIC and staging areas), some methods can be rewritten to make them more efficient/logical

### CODE REVIEW

File: tools/directory/src/molgenis_emx2/directory_client/categories.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    AUTOIMMUNE = "autoimmune"
+    CARDIOVASCULAR = "cardiovascular"
+    COVID19 = "covid19"
+    INFECTIOUS = "infectious"
+    METABOLIC = "metabolic"
+    NERVOUS_SYSTEM = "nervous_system"
+    ONCOLOGY = "oncology"
+    PAEDIATRICS = "paediatrics"
+    #    PAEDIATRIC = "paediatric_only"
+    #    PAEDIATRIC_INCLUDED = "paediatric_included"
```
Comment (esthervanenckevort) [inline tools/directory/src/molgenis_emx2/directory_client/categories.py:22]: Remove commented out code

File: tools/directory/src/molgenis_emx2/directory_client/directory_client.py [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+    ONTOLOGY_SCHEMA = "DirectoryOntologies"
+
+    def get(
+        self,
+        table: str,
+        query_filter: str = None,
+        schema: str = None,
+        as_df: bool = False,
+    ) -> list | pd.DataFrame:
+        # response_data = super().get(table, query_filter, schema, as_df)
```
Comment (esthervanenckevort) [inline tools/directory/src/molgenis_emx2/directory_client/directory_client.py:68]: Remove commented out code

File: tools/directory/src/molgenis_emx2/directory_client/directory_client.py [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        :param matching_attrs: a list with the relevant level or levels of the
+                               "matching" ontology code columns
+        :return: an OntologyTable
+        """
+        matching_attrs = matching_attrs if matching_attrs else []
+
+        rows = self.get(
+            schema=self.ONTOLOGY_SCHEMA,
+            table=entity_type_id,
+            #   attributes=f"id,{parent_attr},ontology,{','.join(matching_attrs)}",
```
Comment (esthervanenckevort) [inline tools/directory/src/molgenis_emx2/directory_client/directory_client.py:158]: Remove commented out code

File: tools/directory/src/molgenis_emx2/directory_client/directory_client.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+
+    def get_quality_info(self) -> QualityInfo:
+        """
+        Retrieves the quality information identifiers for biobanks and collections.
+        :return: a QualityInfo object
+        """
+
+        biobank_qualities = self.get(
+            table="QualityInfoBiobanks",
+            #      attributes="id,biobank,assess_level_bio",
```
Comment (esthervanenckevort) [inline tools/directory/src/molgenis_emx2/directory_client/directory_client.py:176]: Remove commented out code

File: tools/directory/src/molgenis_emx2/directory_client/directory_client.py [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+        )
+        collection_qualities = self.get(
+            table="QualityInfoCollections",
+            #    attributes="id,collection,assess_level_col",
```
Comment (esthervanenckevort) [inline tools/directory/src/molgenis_emx2/directory_client/directory_client.py:180]: Remove commented out code

File: tools/directory/src/molgenis_emx2/directory_client/directory_client.py [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+            self._validate_codes(codes, nodes)
+        return self._to_nodes(nodes)
+
+    def get_external_node(self, code: str) -> ExternalServerNode:
+        """
+        Retrieves a single ExternalServerNode object from the national nodes table.
+        :param code: node to get by code
+        :return: ExternalServerNode object
+        """
+        # Query filter on null is not yet possible therefore this workaround
```
Comment (esthervanenckevort) [inline tools/directory/src/molgenis_emx2/directory_client/directory_client.py:233]: If there is an issue in Github for this, please reference it in the comment

**Review by esthervanenckevort — COMMENTED** [review]
De code ziet er goed gestructureerd uit. Ik zag nog een paar plekken waar code in commentaar stond. Dat zou ik nog even verwijderen.
Ik heb de code niet inhoudelijk getest, dat laat ik aan Hessel over. Als hij de PR goedkeurt kun je mergen.

**Review by hslh — APPROVED** [review]
Looks good, as far as I can tell. I tested it on my own testserver, with an external node on the bbmri-testserver, and only with a dummy PID service and it worked well. I did have to do some extra work to get it to run, I don't know to what extent we want to include those in the documentation somewhere:

- create the .env, it would be helpful to have an example .env included
- add rows to the NationalNodes table in the target directory (1 local, CZ, 1 EU and 1 external, NL)
- make sure the staging areas followed the naming conventions of prefix + national node code 
- install python-dotenv, which isn't listed as a requirement
- empty the 'studies'-column in the Collections tables of the staging areas

**Pull request merged**